src/auth/auth_gitlab.py

```diff
@@ -0,0 +1,339 @@
+                    self.user_states[userData]['updated'] = now - self.gitlab_update - 1
+            LOGGER.info("Readed state from file %s: " % self.gitlab_dump + str(self.user_states))
+
+        self.gitlab_group_search = params_dict.get('group_search')
+
+        self._client_visible_config['client_id'] = self.client_id
+        self._client_visible_config['oauth_url'] = self._OAUTH_AUTHORIZE_URL % self._GITLAB_PREFIX
+        self._client_visible_config['oauth_scope'] = 'api' if self.gitlab_group_support else 'read_user'
+
+    def authenticate(self, request_handler):
+        code = request_handler.get_argument('code', False)
+
+        if not code:
+            LOGGER.error('Code is not specified')
+            raise AuthBadRequestException('Missing authorization information. Please contact your administrator')
+
+        return self.validate_user(code, request_handler)
+
+    def is_active(self, user, request_handler):
+        access_token = request_handler.get_secure_cookie('token')
+        if access_token is None:
+            return False
+        access_token = access_token.decode("utf-8")
+
+        if self.user_states.get(user) is None:
+            LOGGER.debug("User %s not found in state" % user)
+            return False
+
+        if self.user_states[user]['state'] is None or self.user_states[user]['state'] != "active":
+            LOGGER.info("User %s state inactive: " % user + str(self.user_states[user]))
+            del self.user_states[user]
+            self.dump_sessions_to_file()
+            return False
+
+        now = time.time()
+        # check session ttl
+        if self.session_expire and (self.user_states[user]['visit'] + self.session_expire) < now:
+            del self.user_states[user]
+            LOGGER.info("User %s session expired, logged out" % user)
+            self.dump_sessions_to_file()
+            return False
+
+        self.user_states[user]['visit'] = now
+
+        # check gitlab response ttl, also check for stale updating (ttl*2)
+        if self.gitlab_update is not None:
+            stale = (self.user_states[user]['updated'] + max(self.gitlab_update*2, 60)) < now
+            ttl_expired = (self.user_states[user]['updated'] + self.gitlab_update) < now
+            updating_now = self.user_states[user]['updating'] is True
+            if ttl_expired and (not updating_now or stale):
+                if self.gitlab_group_support:
+                    self.do_update_groups(user, access_token)
+                else:
+                    self.do_update_user(user, access_token)
+
+        return True
+
+    def get_groups(self, user, known_groups=None):
+        if self.user_states.get(user) is None:
+            return []
+        if self.user_states[user]['groups'] is None:
+            return []
+        return self.user_states[user]['groups']
+
+    def logout(self, user, request_handler):
+        request_handler.clear_cookie('token')
+
+    def clean_expired_sessions(self):
```

Comment:
Maybe you can call this guy from some timer?

Reply:
May be, but it's really hard for me to find how to set timer in tornado event loop. I am new in python and tornado :) And there is no big advantages in timer. Just free some memory but size of `user_states` never be more than total number of users. And you should be prepared for the situation that all users online at a time.